python/dgl/sampling/randomwalks.py

```diff
@@ -169,8 +169,17 @@ def random_walk(g, nodes, *, metapath=None, length=None, prob=None, restart_prob
         metapath = [g.get_etype_id(etype) for etype in metapath]
 
     gidx = g._graph
-    nodes = F.to_dgl_nd(utils.prepare_tensor(g, nodes, 'nodes'))
-    metapath = F.to_dgl_nd(utils.prepare_tensor(g, metapath, 'metapath'))
+    nodes = utils.prepare_tensor(g, nodes, 'nodes')
+    assert F.max(nodes, dim=0) < g.num_nodes(g.canonical_etypes[metapath[0]][0]), \
+        "Seed node ID exceeds the maximum number of nodes."
+    metapath = utils.prepare_tensor(g, metapath, 'metapath')
+    if g.is_pinned():
+        assert prob is None, "Prob must be None when the graph is pinned."
+        assert F.device_type(F.context(nodes)) != 'cpu', \
+            "Seed nodes must be on GPU when the graph is pinned."
+        metapath = F.copy_to(metapath, F.context(nodes))
```

Comment:
For L176-L178, can we check them in C++?

Reply:
Moved `prob` check to C++.

For the device check of `nodes`, I suggest keeping it on the Python side. The random walk kernel dispatching is based on the context of `nodes` now. In theory, passing `nodes` on the CPU to a pinned graph should work as well but may have unexpected behavior, for example, sampling is done by CPU and the results are also on CPU memory, which may be hard to realize for users.

Reply:
Resolved.